File: Practice/AllDomains/Tutorials/30DaysOfCode/Day23BSTLevel-OrderTraversal.cs
/*	author:@shivkrthakur  */
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
class Node{
    public Node left,right;
    public int data;
    public Node(int data){
        this.data=data;
        left=right=null;
    }
}
class Solution{
	static void levelOrder(Node root)
	{
        Queue<Node> queue = new Queue<Node>();
  		//Write your code here
        if(root != null) queue.Enqueue(root);
        
        while(queue.Count > 0)
        {
            Node node = queue.Dequeue();
            Console.Write(node.data);
            Console.Write(" ");
            if(node.left != null) queue.Enqueue(node.left);
            if(node.right != null) queue.Enqueue(node.right);
        }
    }

	static Node insert(Node root, int data)
	{
		if(root==null){
			return new Node(data);
		}
		else{
			Node cur;
			if(data<=root.data){
				cur=insert(root.left,data);
				root.left=cur;
			}
			else{
				cur=insert(root.right,data);
				root.right=cur;
			}
			return root;
		}
    }
	
    static void Main(String[] args)
	{
        Node root=null;
        int T=Int32.Parse(Console.ReadLine());
        while(T-->0){
            int data=Int32.Parse(Console.ReadLine());
            root=insert(root,data);            
        }
        levelOrder(root);
    }
}

File: Practice/AllDomains/Tutorials/30DaysOfCode/Day23BSTLevel-OrderTraversal.js
/*	author:@shivkrthakur  */
function Node(data) {
    this.data = data;
    this.left = null;
    this.right = null;
}; // End of function Node

// Start of function BinarySearchTree
function BinarySearchTree() {
    this.insert = function(root, data) {
        if (root === null) {
            this.root = new Node(data);

            return this.root;
        }

        if (data <= root.data) {
            if (root.left) {
                this.insert(root.left, data);
            } else {
                root.left = new Node(data);
            }
        } else {
            if (root.right) {
                this.insert(root.right, data);
            } else {
                root.right = new Node(data);
            }
        }

        return this.root;
    };

    // Start of function levelOrder
    this.levelOrder = function(root) {        // Add your code here
        // To print values separated by spaces use process.stdout.write(someValue + ' ')
        var queue = [];
        var output = [];
        if(root){
            queue.push(root);
            while(queue.length > 0){
                root = queue.shift();
                if(root.left) queue.push(root.left);
                if(root.right) queue.push(root.right);
                //console.log(root.data);
                output.push(root.data);
            }
        }
        console.log(output.join(' '));
    }; // End of function levelOrder
}; // End of function BinarySearchTree

process.stdin.resume();
process.stdin.setEncoding('ascii');

var _input = "";

process.stdin.on('data', function (data) {
  _input += data;
});

process.stdin.on('end', function () {
  var tree = new BinarySearchTree();
  var root = null;

  var values = _input.split('\n').map(Number);

  for (var i = 1; i < values.length; i++) {
      root = tree.insert(root, values[i]);
  }

  tree.levelOrder(root);
});
